dev-docs/design/how-lit-html-works.md

```diff
@@ -0,0 +1,426 @@
+
+A value of `null`, `undefined`, or `nothing` will remove any previously set listeners without adding a new listener.
+
+#### ElementPart.\_$setValue
+
+Nothing can be committed via an `ElementPart`. Instead `resolveDirective` is called, allowing directives to hook into the `ElementPart` position.
+
+For example, [`@lit-labs/motion` has an `animate` directive](https://lit.dev/playground/#sample=examples/motion-simple) that can be applied on an element in the `ElementPart` location to then manage the element's animations.
+
+# Efficient Updates
+
+So far we've detailed what happens on the first render: `render(counterUi(0), container)` in the example, but what happens when the "Increment" button is clicked and `render(counterUi(1), container)` is invoked?
+
+The `render()` function looks for a `'_$litPart$'` property on the container and finds a `ChildPart`. `_$setValue` is called on the container's `ChildPart` passing the `TemplateResult` returned from `counterUi(1)`.
+
+The `TemplateResult` for `counterUi(1)` is:
+
+```js
+{
+  _$litType$: 1,
+  strings: [
+    '<span class="',
+    '">',
+    '</span><button @click=',
+    '>Increment</button>'
+  ],
+  values: [
+    "odd",
+    1,
+    () => render(counterUi(2), container)
+  ]
+}
+```
+
+Instead of redoing the work in the **Prepare** phase, the Lit `Template` is fetched from the cache using the template's `strings` Template Strings Array reference which is stable across multiple tag function evaluations as described in [Foundational Pieces](#foundational-pieces).
+
+The container's `ChildPart` stores the last committed value — which in this case is a `TemplateInstance` referencing the `Template` returned from the cache. Because we're rendering the same `Template` to this `ChildPart` the **Create** phase is skipped.
+
+All that is left is to iterate over the previously created `Part`s and call `_$setValue` with the new values. For demonstrative purposes, inlined this looks like:
+
+```js
+AttributePart._$setValue('odd');
+ChildPart._$setValue(1);
+EventPart._$setValue(() => render(counterUi(2), container));
+```
+
+This results in the DOM updating with a green underlined number one, and the event listener is swapped with the new listener.
+
+If any of the values being set remain the same (by triple equal check), Lit will not do any DOM
+operations.
+
+# Future work
+
+lit-html is fast, but what if it could be faster!?
+
+- [Declarative syntax for creating DOM Parts](https://github.com/WICG/webcomponents/issues/990). This could remove the requirement for lit-html to walk the `<template>` element's tree and manually instantiate Parts. This reduces the tree walk done in the **Create** phase.
+- Precompiling lit-html `html` tag functions (tracked in https://github.com/lit/lit/issues/189). This would allow opting into doing the **Prepare** phase in a compile step done by build tooling. Runtime would no longer have a **Prepare** phase.
```

Comment:
Woah - the Template Instantiation proposal looks like it would remove a lot (or all) of the book keeping we need to do with our Parts. Should I add a dot point like:

```md
[Template instantiation](https://github.com/WICG/webcomponents/blob/gh-pages/proposals/Template-Instantiation.md#html-template-instantiation) could also greatly reduce prepare phase work and builds on top of the DOM parts proposal.
```

It looks like it would reduce the work we do in the `Create` and `Update` phases, by giving us a direct way to `update` a TemplateInstance?

Reply:
I think it's fair to say that it would let us do more with less code because the browser would handle many of the steps, but we'd still need code for directives, and it's an open question how we'd handle like `nothing`. I expect `Create` would be faster, but `Update` might be about as fast (though smaller)

Reply:
Added dot point! Thank you